tokio/src/net/windows/named_pipe.rs

```diff
@@ -0,0 +1,388 @@
+//! Windows named pipes.
+
+use mio::windows::NamedPipe as MioNamedPipe;
+use miow::pipe::{NamedPipe as RawNamedPipe, NamedPipeBuilder};
+use winapi::{
+    shared::winerror::*,
+    um::{minwinbase::*, namedpipeapi::WaitNamedPipeW, winbase::*},
+};
+
+use std::{
+    ffi::{OsStr, OsString},
+    fs::OpenOptions,
+    future::Future,
+    io::{self, ErrorKind, IoSlice, Result},
+    mem,
+    os::windows::prelude::*,
+    pin::Pin,
+    sync::Mutex,
+    task::{Context, Poll},
+};
+
+use crate::io::{AsyncRead, AsyncWrite, Interest, PollEvented, ReadBuf, Ready};
+
+/// Default in/out buffer size.
+pub const DEFAULT_BUFFER_SIZE: u32 = 65_536;
+
+fn mio_from_miow(pipe: RawNamedPipe) -> MioNamedPipe {
+    // Safety: nothing actually unsafe about this. The trait fn includes `unsafe`.
+    unsafe { MioNamedPipe::from_raw_handle(pipe.into_raw_handle()) }
+}
+
+/// Connecting instance future.
+#[derive(Debug)]
+enum ConnectingInstance {
+    New(NamedPipe),
+    Connecting(NamedPipe),
+    Error(io::Error),
+    Ready(Option<NamedPipe>),
+}
+
+impl ConnectingInstance {
+    fn new(mio: MioNamedPipe) -> Result<Self> {
+        Ok(Self::New(NamedPipe::server(mio)?))
+    }
+}
+
+impl Future for ConnectingInstance {
+    type Output = Result<NamedPipe>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        match mem::replace(&mut *self, ConnectingInstance::Ready(None)) {
+            Self::Ready(None) => {
+                // poll on completed future
+                Poll::Pending
+            }
+            Self::Ready(Some(pipe)) => Poll::Ready(Ok(pipe)),
+            Self::Connecting(pipe) => match pipe.poll_write_ready(cx) {
+                Poll::Ready(Ok(_)) => Poll::Ready(Ok(pipe)),
+                Poll::Ready(Err(err)) => Poll::Ready(Err(err)),
+                Poll::Pending => {
+                    *self = Self::Connecting(pipe);
+                    Poll::Pending
+                }
+            },
+            Self::New(pipe) => match pipe.io_ref().connect() {
+                Ok(()) => {
+                    *self = Self::Connecting(pipe);
+                    Poll::Pending
+                }
+                Err(err) if err.kind() == ErrorKind::WouldBlock => {
+                    *self = Self::Connecting(pipe);
+                    Poll::Pending
```

Comment:
Unless I am misunderstanding the implementation, would we not need to arrange for another poll here if this `Future` implementation is supposed to work with `await`? Otherwise, `poll` will never be called again and we will never make progress?

Reply:
Hmm, wouldn't it be notified from [connect_done][1]?

[1]: https://github.com/tokio-rs/mio/blob/master/src/sys/windows/named_pipe.rs#L584

Reply:
I don't think so because `cx` from this `poll` is not sent in to connect so how would it wake?

Reply:
I haven't looked into this particular case further, but the way you usually register the waker (see e.g. the `AsyncFd` example) is by having a loop that goes back and polls for readiness again when you get a `WouldBlock`.

Reply:
Yeah, that seems appropriate. Wouldn't it also be possible to implement `accept` in terms of `async_io()` in `Registration` just like `accept` in `UnixListener`?

Reply:
> When the server is a separate executable, accept() never returns

@abbec, this should be fixed by 5103eb3, thanks.